include/aws/s3/s3_client.h

```diff
@@ -228,6 +228,12 @@ struct aws_s3_meta_request_options {
     /* Callback for when the meta request has completely cleaned up. */
     aws_s3_meta_request_shutdown_fn *shutdown_callback;
 
+    /* Support meta requests over HTTP and HTTPS */
+    enum aws_s3_meta_request_tls_mode tls_mode;
```

Comment:
Do we need this tls_mode override the configuration for the whole client?

Reply:
Okay, I read thought the issue again. Well, I think we should move the logic of taking an URL to the C part instead. So, the java will just create a C URL from java URL and pass in here. You can use the structure [here](https://github.com/awslabs/aws-c-io/blob/main/include/aws/io/uri.h#L15)

Reply:
@TingDaoK Thank you. Before I begin work on the changes, I would like to make sure I'm clear on your intent.

Presently (before my patch) the `S3Client.java` / `makeMetaRequest` method receives an instance of `software.amazon.awssdk.crt.http.HttpRequest` containing a host header. The `HttpRequest` is passed to the JNI as a struct, and the host name in the request header eventually makes its way down to the `s_s3_endpoint_create_http_connection_manager` function.

If I understand - your direction for this PR is to have the `S3Client.java` / `makeMetaRequest` method accept a URI string. The Java method would pass that string to the JNI `...s3ClientMakeMetaRequest` function, which would parse the string into the struct you referenced, and pass that struct down the call chain where it would be used instead of the http request host header. At some point, a decision needs to be made how far down the call chain to pass that struct intact vs. picking it apart passing the pieces as function args. But I can work through that. (Probaby the struct should go all the way to the `s_s3_endpoint_create_http_connection_manager` function and replace any duplicative args along the way.)

Please let me know if I captured your intent. Thank you.